### ISSUE TITLE
Item with ItemHandler capability and custom getShareTag() and readShareTag() methods gets reset when opening creative inventory on dedicated server

### ISSUE THREAD

**Issue by vemerion** [MinecraftForge/MinecraftForge#7359]
**Minecraft Version:** 1.15.2

**Forge Version:** 31.2.0

**Steps to Reproduce:**
 1. Create an Item with an ItemHandler capability and custom getShareTag() and readShareTag() methods. (for example TheGreyGhost's MinecraftByExample MBE32 demonstrates this bug (https://github.com/TheGreyGhost/MinecraftByExample))
 2. Start and join a dedicated server
 3. Add some ItemStack to your custom item
4. Open the creative inventory
5. The ItemHandler for your custom Item should now have been erroneously reset

**Description of issue:**
It seems like the content of custom items with capabilities gets reset when you open the creative inventory while on a dedicated server. When I went looking through the forge source code I found this line in the writeItemStack(ItemStack, boolean) method in the PacketBuffer class:

``compoundnbt = limitedTag ? stack.getShareTag() : stack.getTag();``

It seems to me that the expression should be flipped, or am I way off?

**Comment by brandon3055** [comment]
This issue is still present in 1.16.4 Forge version: 35.1.13

'The root issue here has nothing to do with shareTag and ItemHandlers specifically and is purely related to the lack of capability syncronization when in the creative inventory, as such the title should be "Capability data wiped client and server side when opening Creative inventory on the dedicated server"'

This is an issue that from what i can tell goes all the way back to 2016 #2523 and it makes capability data completely useless on itemstacks.

The problem is that the Creative menu is entirely client side. Unlike every other inventory in existence that detects changes server side and syncs them to the client. The creative inventory works the exact opposite way. Changes are detected client side and then the client side items are sent to the server. The problem is that the method used to send items to the server `PacketBuffer.writeItemStack` and `PacketBuffer.readItemStack` does not send capability data so the moment you open the creative inventory everything is synced to the server and in the process all capability data is lost. 

Unfortunately there is really no clean way to fix this that wouldn't break compatibility with vanilla. 
Something like this would fix the issue and maintain compatibility with vanilla but this is a nasty hack. 
This was written by @covers1624 as a "quick test" to see if it would fix the issue and it does.

`PacketBuffer.java`
```java
   public PacketBuffer writeItemStack(ItemStack stack, boolean limitedTag) {
      if (stack.isEmpty()) {
         this.writeBoolean(false);
      } else {
         this.writeBoolean(true);
         Item item = stack.getItem();
         this.writeVarInt(Item.getIdFromItem(item));
         this.writeByte(stack.getCount());
         CompoundNBT compoundnbt = null;
         if (item.isDamageable(stack) || item.shouldSyncTag()) {
            compoundnbt = limitedTag ? stack.getShareTag() : stack.getTag();
            CompoundNBT caps = stack.serializeNBT().getCompound("ForgeCaps");
            if (!caps.isEmpty()) {
               if (compoundnbt == null) {
                  compoundnbt = new CompoundNBT();
               } else {
                  compoundnbt = compoundnbt.copy();
               }
               compoundnbt.put("__Network_ForgeCaps", caps);
            }
         }

         this.writeCompoundTag(compoundnbt);
      }

      return this;
   }
```
`ItemStack.java`
```java
   public void setTag(@Nullable CompoundNBT nbt) {
      this.tag = nbt;
      if (this.getItem().isDamageable(this)) {
         this.setDamage(this.getDamage());
      }
      if (nbt != null && nbt.contains("__Network_ForgeCaps")) {
         CompoundNBT caps = tag.getCompound("__Network_ForgeCaps");
         tag.remove("__Network_ForgeCaps");
         deserializeCaps(caps);
      }
   }
```

**Comment by brandon3055** [comment]
I have created a PR that allows greater control over the full nbt tag being send over the network. This can be used by mods as a temporary solution for this issue until forge has a proper fix.
#7565

**Comment by alcatrazEscapee** [comment]
I'm curious about the "hacky" fix - it seems to me that the fix just involves always synchronizing capability nbt? i.e. what #6954 did, but in a much more brutish way.

In which case, I would hope the actual solution would be a capability sync implementation as the linked PR, something I've wanted for quite some time as this issue (the creative backwards sync), and the other lacking feature of item stack capability sync (being able to sync capability data on items *not from your mod*, i.e. ones attached to external items) are quite the pain to deal with.

**Comment by brandon3055** [comment]
@alcatrazEscapee thing is you don't *really* need a capability share tag. Sure it would be very nice if capabilities were synced by default but modders can already work around the issue using the existing share tag like so. [IModularItem.java](https://github.com/brandon3055/Draconic-Evolution/blob/4af607da1f7eb144cd6fed5708611a86356f5c66/src/main/java/com/brandon3055/draconicevolution/items/equipment/IModularItem.java#L219-L227) [DECapabilities.java](https://github.com/brandon3055/Draconic-Evolution/blob/28e310fc0f5a0f8a19d4370470f7c2afa49a499b/src/main/java/com/brandon3055/draconicevolution/api/capability/DECapabilities.java#L31-L103)

The creative reverse synchronization is really a separate issue related to the fact the ItemStack packet does not include capability data when doing a full sync. Giving the item control over what data is send in a 'full sync' as i have done in #7565 means that the same code used for syncing via share tag (see above) could also apply to the full tag. The idea behind my pr was to add a simple bare minimum patch that can be used to fix issues like this.

**Comment by alcatrazEscapee** [comment]
> thing is you don't really need a capability share tag.

@brandon3055 Modders can work around the issue for their own items, yes, but insofar I'm not aware of a solution for syncing capabilities that have been attached to external items (i.e. via `AttachCapabilitiesEvent`). `getShareTag` is not sufficient for this purpose. That was what my question was directed towards, as that has been a pain point for me for a while.

(I'm aware this isn't the same issue as the creative inventory, of which I've also been foiled by, but it's not completely solved yet.)

**Comment by brandon3055** [comment]
You do have a good point there. I did not consider capabilities attached via events.

**Comment by LexManos** [comment]
So if Im understanding the issue correctly, the problem is that the creative GUI treats the entire inventory as if it should be client side controlled.
Is there a possibility of making that not the case. Where it doesn't sync the items that are inside the 'inventory' slots and not the creative GUI. 

I'm not a fan of the PR because it adds yet another modder facing thing that could make things weird.
The best option would be to have the  client not sync the inventory slots to the server in the creative gui.
Except for the newly added items, which in turn should automatically sync the FULL nbt data C->S.
The other option is to make Forge aware that you're opening a creative GUI and force syncing all the existing inventory items in their entirety S->C as it loads. But thats a lot of data which is why the share tags exist at all.

Basically, this needs more work, and I want it done in a way that doesn't expose hacky work arounds that modders have to care about.
